Spring boot security

denysokhrymovych opened the pull request:
add required DTOs and mapper,
add AuthController,
add FieldMatch annotation,
handle the RegistrationException using the GlobalExceptionHandler

denysokhrymovych committed:
add User entity,
add required DTOs and mapper,
add AuthController,
add FieldMatch annotation,
handle the RegistrationException using the GlobalExceptionHandler

bhdnchui commented on src/main/java/com/example/bookshop/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+    private String password;
+    private String repeatPassword;
+
+    @NotNull(message = "First name cannot be null")
```
Let's use `@NotBlank` for better validation

bhdnchui commented on src/main/java/com/example/bookshop/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+    @Size(max = 255, message = "First name cannot be greater than 255 characters")
+    private String firstName;
+
+    @NotNull(message = "Last name cannot be null")
```
Let's use `@NotBlank` for better validation

bhdnchui commented on src/main/java/com/example/bookshop/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+    @Email
+    private String email;
+
+    private String password;
+    private String repeatPassword;
```
Add constraints for these fields.

bhdnchui commented on src/main/java/com/example/bookshop/validation/PasswordMatchesValidator.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public boolean isValid(Object obj, ConstraintValidatorContext constraintValidatorContext) {
+        UserRegistrationRequestDto userDto = (UserRegistrationRequestDto) obj;
```
Let's avoid class casting.

bhdnchui commented on src/main/java/com/example/bookshop/validation/PasswordMatchesValidator.java:
```diff
@@ -0,0 +1,17 @@
+    @Override
+    public boolean isValid(Object obj, ConstraintValidatorContext constraintValidatorContext) {
+        UserRegistrationRequestDto userDto = (UserRegistrationRequestDto) obj;
+        return userDto.getPassword().equals(userDto.getRepeatPassword());
```
Let's compare passwords with Objects#equals

bhdnchui commented on src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -44,6 +45,15 @@ public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExcept
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<ExceptionDto> handleRegistrationException(RegistrationException ex) {
+        ExceptionDto exceptionDto = new ExceptionDto();
+        exceptionDto.setTimestamp(LocalDateTime.now());
+        exceptionDto.setStatus(HttpStatus.BAD_REQUEST);
+        exceptionDto.setMessage(ex.getMessage());
+        return new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
```
Why do you return ExceptionDto dto here, if you return LinkedHashMap in handleEntityNotFoundException method.
Be consistent. Let's reuse code from handleEntityNotFoundException method to handle RegistrationException

bhdnchui reviewed (changes requested):
Good job! Just a few points to consider.

denysokhrymovych committed:
use @NotBlank,
remove casting from isValid,
use Objects.equals,
use Map in handleRegistrationException

bhdnchui commented on src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -44,6 +44,15 @@ public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExcept
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
This code is pretty equal to handleEntityNotFoundException method body. 
Please, avoid code duplication.

bhdnchui reviewed (changes requested):
Just one point to fix.

denysokhrymovych committed:
avoid code duplication in CustomGlobalExceptionHandler

d1sam commented on src/main/java/com/example/bookshop/config/SecurityConfig.java:
You don't need to add SecurityConfig in this hw

d1sam commented on src/main/java/com/example/bookshop/model/Role.java:
you don't need to add role in this homework

d1sam commented on src/main/java/com/example/bookshop/validation/FieldMatch.java:
```diff
@@ -0,0 +1,19 @@
+@Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface FieldMatch {
```
```suggestion
public @interface PasswordsFieldsMatch {
```

d1sam replied:
because you check only passwords in your `FieldMatchValidator`

d1sam commented on src/main/java/com/example/bookshop/model/User.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
```suggestion
        return !isDeleted;
```

d1sam commented on src/main/java/com/example/bookshop/model/User.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
```suggestion
        return !isDeleted;
```

d1sam commented on src/main/java/com/example/bookshop/model/User.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
```suggestion
        return !isDeleted;
```

d1sam reviewed (changes requested):
Some changes should be done according to me comment

denysokhrymovych committed:
remove SecurityConfig and Role

d1sam reviewed (approved):
Well done!
